woocommerce/woocommerce-ios
# Dynamic Dashboard: Reviews card states and unit tests

Please do not merge until target is trunk

Closes: #12742


## Description


This PR completes the Reviews card task by adding the following:
1. Finalizing UI for loading state,
2. UI for error state,
3. UI for empty state (both for when there's no reviews at all in the site, and when there's no reviews after filtering). The wording here follows [Android's implementation](https://github.com/woocommerce/woocommerce-android/pull/11491/files#diff-b931a16277ac6133750e3d68c243da133842255a53e451802acd40aad3645e81R423-R424).
4. Dismiss and onError retry trackings
5. Unit tests

## Testing instructions


Please check and compare with screenshots posted below:

1. Start with no internet, ensure "error state" is shown.
2. Retry, ensure this is tracked: `Tracked dynamic_dashboard_card_retry_tapped, ...` with `type: reviews` included.
3. Restart test, now with internet on a site with no reviews. ensure "Empty state case: absolutely no reviews" case shown.
4. Add one review on the site, but do not approve it. Go back to site, filter by "Approved". Ensure "Empty state case: no filtered reviews" case is shown.
5. Try reloading and ensure loading state is shown correctly.
6. HIde the card, ensure this is tracked: `Tracked dynamic_dashboard_hide_card_tapped, ...` with `type: reviews` included.

## Screenshots


### Error case:
 <img src="https://github.com/woocommerce/woocommerce-ios/assets/266376/a7d5e960-eeef-44df-b0b1-eec49d5153b2" width="42%"> 

| Empty state case: absolutely no reviews | Empty state case:  no filtered reviews |
|-|-|
|![Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-22 at 09 42 07](https://github.com/woocommerce/woocommerce-ios/assets/266376/66e14eea-d587-4b2c-9e14-e0d046115054) | ![Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-22 at 09 39 49](https://github.com/woocommerce/woocommerce-ios/assets/266376/98613b44-d4a8-48a9-8a8f-a2603368c2fd) |


| Loading state case: no reviews ever loaded | Loading state case: some reviews were shown before |
|-|-|
| ![Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-20 at 16 54 32](https://github.com/woocommerce/woocommerce-ios/assets/266376/de7ef8c3-8984-488d-8fe8-ac52e56e8bf4) | ![Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-20 at 16 54 21](https://github.com/woocommerce/woocommerce-ios/assets/266376/3a37992b-0b51-4f58-bb68-8700e744cc67)  |
